agenthub/codeact_agent/codeact_agent.py

```diff
@@ -93,56 +85,82 @@ def step(self, state: State) -> Action:
             assert state.plan.main_goal, 'Expecting instruction to be set'
             self.messages = [
                 {'role': 'system', 'content': SYSTEM_MESSAGE},
-                {'role': 'user', 'content': state.plan.main_goal},
+                {
+                    'role': 'user',
+                    'content': (
+                        f'Here is an example of how you can interact with the environment for task solving:\n{EXAMPLES}\n\n'
+                        f"NOW, LET'S START!\n\n{state.plan.main_goal}"
+                    )
+                },
             ]
         updated_info = state.updated_info
         if updated_info:
             for prev_action, obs in updated_info:
                 assert isinstance(
-                    prev_action, (CmdRunAction, AgentEchoAction)
-                ), 'Expecting CmdRunAction or AgentEchoAction for Action'
-                if isinstance(
-                    obs, AgentMessageObservation
-                ):  # warning message from itself
+                    prev_action, self.SUPPORTED_ACTIONS
+                ), f'{prev_action.__class__} is not supported (supported: {self.SUPPORTED_ACTIONS})'
+                # prev_action is already added to self.messages when returned
+
+                # handle observations
+                assert isinstance(
+                    obs, self.SUPPORTED_OBSERVATIONS
+                ), f'{obs.__class__} is not supported (supported: {self.SUPPORTED_OBSERVATIONS})'
+                if isinstance(obs, (AgentMessageObservation, UserMessageObservation)):
                     self.messages.append(
                         {'role': 'user', 'content': obs.content})
+
+                    # User wants to exit
+                    if obs.content.strip() == '/exit':
+                        return AgentFinishAction()
                 elif isinstance(obs, CmdOutputObservation):
                     content = 'OBSERVATION:\n' + obs.content
                     content += f'\n[Command {obs.command_id} finished with exit code {obs.exit_code}]]'
                     self.messages.append({'role': 'user', 'content': content})
+                elif isinstance(obs, IPythonRunCellObservation):
+                    content = 'OBSERVATION:\n' + obs.content
+                    self.messages.append({'role': 'user', 'content': content})
                 else:
                     raise NotImplementedError(
                         f'Unknown observation type: {obs.__class__}'
                     )
+
         response = self.llm.completion(
             messages=self.messages,
-            stop=['</execute>'],
+            stop=[
+                '</execute_ipython>',
+                '</execute_bash>',
+                # chatML in case ollama does not stop
```

Comment:
(nit) what's specific about ollama here? Are you referring to weaker LLMs in general here?

Reply:
This was more like an ollama stopword issue that happend while i was testing. it is no long happening, i'm removing it. Thanks for the reminder!